src/main/java/com/yahoo/bullet/querying/partitioning/SimpleEqualityPartitioner.java

```diff
@@ -0,0 +1,215 @@
+        }
+        return new ArrayList<>(keys);
+    }
+
+    private static boolean hasNonANDLogicals(List<Clause> filters) {
+        return filters != null && filters.stream().anyMatch(SimpleEqualityPartitioner::hasNonANDLogicals);
+    }
+
+    private static boolean hasNonANDLogicals(Clause clause) {
+        if (clause instanceof FilterClause) {
+            return false;
+        }
+        return clause.getOperation() != Clause.Operation.AND || hasNonANDLogicals(((LogicalClause) clause).getClauses());
+    }
+
+    private void mapFieldToFilters(Clause clause, Map<String, List<FilterClause>> mapping) {
+        if (clause instanceof FilterClause) {
+            mapFieldToFilter((FilterClause) clause, mapping);
+            return;
+        }
+        List<Clause> clauses = ((LogicalClause) clause).getClauses();
+        // Cannot have null non-AND logicals as it is checked for.
+        clauses.forEach(c -> this.mapFieldToFilters(c, mapping));
+    }
+
+    private void mapFieldToFilter(FilterClause clause, Map<String, List<FilterClause>> mapping) {
+        String field = clause.getField();
+        if (clause.getOperation() != Clause.Operation.EQUALS || !fieldSet.contains(field)) {
+            return;
+        }
+        List<FilterClause> list = mapping.getOrDefault(field, new ArrayList<>());
+        list.add(clause);
+        mapping.put(field, list);
+    }
+
+    private boolean hasInvalidFilterClauses(List<FilterClause> filters) {
+        if (filters == null || filters.size() != 1)  {
+            return true;
+        }
+        FilterClause filter = filters.get(0);
+        List values = filter.getValues();
+        return values == null || values.size() != 1;
+    }
+
+    private String getFilterValue(List<FilterClause> singletonFilters) {
+        if (singletonFilters == null) {
+            return NO_FIELD;
+        }
+        // Otherwise, it's a list of size 1 with a singular value, which has been already validated
+        FilterClause filter = singletonFilters.get(0);
+        Object value = filter.getValues().get(0);
+        return makeKeyEntry(filter.getValue(value));
+    }
+
+    private Map<String, String> getFieldValues(BulletRecord record) {
+        Map<String, String> fieldValues = new HashMap<>();
+        for (String field : fields) {
+            Object value = record.extractField(field);
+            fieldValues.put(field, value == null ? NO_FIELD : makeKeyEntry(value.toString()));
+        }
+        return fieldValues;
+    }
+
+    private String binaryToKey(String binary, Map<String, String> values) {
+        // If binary is 010 and fields is [A, B.c, D], the key is [null, values[B.c], null].join(delimiter)
+        int fieldCount = fields.size();
+        String[] keyParts = new String[fieldCount];
+        for (int i = 0; i < fieldCount; ++i) {
+            boolean include = binary.charAt(i) != FALSE_CHAR;
+            keyParts[i] = include ? values.get(fields.get(i)) : NO_FIELD;
+        }
```

Comment:
Use Stream.of(binary. toCharArray).map(char -> xxx).collect(...) ?

Reply:
Used a stream but different since using ints